core/src/upickle/core/JsonPointerVisitor.scala

```diff
@@ -0,0 +1,189 @@
+package upickle.core
+
+import upickle.core.JsonPointerVisitor._
+
+import scala.annotation.tailrec
+import scala.util.control.{NoStackTrace, NonFatal}
+
+/**
+  * Adds a suppressed [[JsonPointerException]] to any thrown exceptions.
+  *
+  * JSON Pointer is standardized by RFC 6901 and commonly used by JSON Schema.
+  *
+  * @see https://tools.ietf.org/html/rfc6901
+  */
+object JsonPointerVisitor {
+
+  def apply[T, J](delegate: Visitor[T, J]): Visitor[T, J] = new JsonPointerVisitor(delegate, RootHasPath)
+
+  /**
+    * JSON Pointer indicating where the problem occurred.
+    * Added as a suppressed exception.
+    *
+    * @param jsonPointer e.g. "/hits/hits/3/_source/foo/bar"
+    * @see https://tools.ietf.org/html/rfc6901
+    */
+  class JsonPointerException(val jsonPointer: String) extends Exception(jsonPointer) with NoStackTrace {
+
+    override def toString: String = jsonPointer
+  }
+
+  /**
+    * Internally, the paths form a linked list back to the root by the visitors themselves.
+    * Compared to something like a List[String] or List[Object], this does not require
+    * extra String allocation or boxing unless we actually ask for the path.
+    */
+  private trait HasPath {
+
+    /**
+      * Forms a chain toward the root.
+      */
+    def parent: Option[HasPath]
+
+    /**
+      * @return name of a single level, if any, e.g. "foo"
+      */
+    def pathComponent: Option[String]
+
+    /**
+      * @return root-to-leaf ordered chain as a List[HasPath]
+      */
+    private def components: List[HasPath] = {
+      // Used rarely. Stack safety > memory efficiency here.
+      @tailrec def listPath(o: Option[HasPath], list: List[HasPath]): List[HasPath] = {
+        o match {
+          case Some(p) => listPath(p.parent, p :: list)
+          case None => list
+        }
+      }
+      listPath(parent, List(this))
+    }
+
+    /**
+      * @return the full path, slash-delimited.
+      */
+    def path: String = components.iterator.map(_.pathComponent).flatten.mkString("/")
+
+    override def toString: String = path
+
+    def wrap[T](f: => T): T = {
+      try {
+        f
+      } catch {
+        case NonFatal(cause) =>
+          cause.addSuppressed(new JsonPointerException(path))
```

Comment:
This would be prettier unsuppressed, but I didn't want to update all the tests or break anyone out there that's catching an `Abort`.